src/Network/Mattermost/Types/Internal.hs

```diff
@@ -99,11 +92,13 @@ data ConnectionData
   , cdConnectionPool :: Pool MMConn
   , cdConnectionCtx  :: C.ConnectionContext
   , cdToken          :: Maybe Token
-  , cdBasicAuth      :: Maybe BasicAuth
+  , cdConnReqTrans   :: HTTP.Request_String -> HTTP.Request_String
+  , cdWsReqTrans     :: WS.Headers -> WS.Headers
```

Comment:
I can see how having these transformations separated is important, but I'm also concerned about this API making it too easy to forget to specify them both when a transformation needs to be applied to both the websocket headers as well as to normal requests. In this patch, `cdConnReqTrans` is only ever used with `populateAuth`, which just modifies headers. So now I wonder: would it make more sense to just combine these into one transformation of type `Headers -> Headers` to ensure that the same transformation gets applied to all kinds of requests? That would be less error-prone. (I'm assuming that the `Headers` type used by `WS` is the same as the one inside the `Request_String`.)

Reply:
Pitily they are not. 
WS.Headers is `type Headers = [(CI ByteString, ByteString)] `
HTTP.Headers is `data Header = Header HeaderName String`

I tried something like this first:
```
data MMReq = 
  HttpReq Request_string
  | WS.Headers
```
but the function `MMReq -> MMReq` could return WS.Headers when pass HTTP.Header.

Reply:
Okay, thanks - that's too bad. If these are different types, there isn't much we can do about that. But then I wonder if a less error-prone API would be to put these two transformations together in a new record and make the user provide them as a pair; this at least makes it hard to forget to provide one when you provide the other.

Reply:
> to put these two transformations together in a new record

Sounds good to me. I changed that.